dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp

```diff
@@ -53,138 +54,212 @@ extern const int ILLFORMAT_RAFT_ROW;
-        if (need_decode)
-            storage->releaseDecodingBlock(block_decoding_schema_epoch, std::move(block_ptr));
-
+    /// Read region data as block.
+    Stopwatch watch;
+    Int64 block_decoding_schema_epoch = -1;
+    BlockUPtr block_ptr = nullptr;
+    bool should_handle_version_col = true;
+    if constexpr (std::is_same_v<ReadList, RegionUncommittedDataList>)
+    {
+        should_handle_version_col = false;
+    }
+    if (need_decode)
+    {
         LOG_TRACE(
             log,
-            "keyspace={} table_id={} region_id={} cost [region decode {}, write part {}] ms",
-            keyspace_id,
-            table_id,
-            region->id(),
-            region_decode_cost,
-            write_part_cost);
-        return true;
-    };
+            "begin to decode keyspace={} table_id={} region_id={}",
+            rw_ctx.keyspace_id,
+            rw_ctx.table_id,
+            region->id());
+        DecodingStorageSchemaSnapshotConstPtr decoding_schema_snapshot;
+        std::tie(decoding_schema_snapshot, block_ptr)
+            = storage->getSchemaSnapshotAndBlockForDecoding(lock, true, should_handle_version_col);
+        block_decoding_schema_epoch = decoding_schema_snapshot->decoding_schema_epoch;
+
+        auto reader = RegionBlockReader(decoding_schema_snapshot);
+        if (!reader.read(*block_ptr, data_list_read, force_decode))
+            return false;
+        rw_ctx.region_decode_cost = watch.elapsedMilliseconds();
+        GET_METRIC(tiflash_raft_write_data_to_storage_duration_seconds, type_decode)
+            .Observe(rw_ctx.region_decode_cost / 1000.0);
+    }
+    if constexpr (std::is_same_v<ReadList, RegionDataReadInfoList>)
+    {
+        writeCommittedBlockDataIntoStorage(
+            rw_ctx,
+            lock,
+            storage,
+            need_decode,
+            block_decoding_schema_epoch,
+            block_ptr,
+            block);
+    }
+    else
+    {
+        // TODO Implement spill logic.
+        RUNTIME_CHECK(false);
+    }
+    LOG_TRACE(
+        log,
+        "keyspace={} table_id={} region_id={} cost [region decode {}, write part {}] ms",
+        rw_ctx.keyspace_id,
+        rw_ctx.table_id,
+        region->id(),
+        rw_ctx.region_decode_cost,
+        rw_ctx.write_part_cost);
+    return true;
+}
+
+template DM::WriteResult writeRegionDataToStorage<RegionUncommittedDataList>(
+    Context & context,
+    const RegionPtrWithBlock & region,
+    RegionUncommittedDataList & data_list_read,
+    const LoggerPtr & log);
+template DM::WriteResult writeRegionDataToStorage<RegionDataReadInfoList>(
+    Context & context,
+    const RegionPtrWithBlock & region,
+    RegionDataReadInfoList & data_list_read,
+    const LoggerPtr & log);
+
+// TODO rename it after we support spill.
+// ReadList could be RegionDataReadInfoList
+template <typename ReadList>
+DM::WriteResult writeRegionDataToStorage(
```

Comment:
The function should later be renamed to something like `writeRegionDataToLowerLayer` or something more suitable.